File: sdk/python/src/panproto/_schema.py
"""Fluent schema builder API.

Builders are immutable: each method returns a new builder instance.
Call :meth:`SchemaBuilder.build` to validate and produce the final schema.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, final

from ._errors import SchemaValidationError
from ._msgpack import pack_schema_ops
from ._wasm import WasmHandle, WasmModule, create_handle

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence

    from ._types import (
        Constraint,
        Edge,
        EdgeOptions,
        HyperEdge,
        SchemaData,
        SchemaOp,
        Vertex,
        VertexOptions,
    )

# ---------------------------------------------------------------------------
# SchemaBuilder
# ---------------------------------------------------------------------------


@final
class SchemaBuilder:
    """Immutable fluent builder for constructing panproto schemas.

    Each mutation method returns a *new* :class:`SchemaBuilder` instance,
    leaving the original unchanged. Operations accumulate and are sent to
    WASM on :meth:`build`.

    Parameters
    ----------
    protocol_name : str
        Name of the owning protocol.
    protocol_handle : WasmHandle
        WASM handle for the registered protocol.
    wasm : WasmModule
        The WASM module used for schema construction.
    ops : Sequence[SchemaOp], optional
        Accumulated schema operations (default ``()``).
    vertices : Mapping[str, Vertex], optional
        Accumulated vertices (default empty).
    edges : Sequence[Edge], optional
        Accumulated edges (default ``()``).
    hyper_edges : Mapping[str, HyperEdge], optional
        Accumulated hyperedges (default empty).
    constraints : Mapping[str, Sequence[Constraint]], optional
        Accumulated constraints keyed by vertex id (default empty).
    required : Mapping[str, Sequence[Edge]], optional
        Accumulated required-edge sets keyed by vertex id (default empty).

    Examples
    --------
    >>> schema = (
    ...     builder
    ...     .vertex("post", "record", VertexOptions(nsid="app.bsky.feed.post"))
    ...     .vertex("post:body", "object")
    ...     .edge("post", "post:body", "record-schema")
    ...     .build()
    ... )
    """

    __slots__ = (
        "_constraints",
        "_edges",
        "_hyper_edges",
        "_ops",
        "_protocol_handle",
        "_protocol_name",
        "_required",
        "_vertices",
        "_wasm",
    )

    def __init__(
        self,
        protocol_name: str,
        protocol_handle: WasmHandle,
        wasm: WasmModule,
        ops: Sequence[SchemaOp] | None = None,
        vertices: Mapping[str, Vertex] | None = None,
        edges: Sequence[Edge] | None = None,
        hyper_edges: Mapping[str, HyperEdge] | None = None,
        constraints: Mapping[str, Sequence[Constraint]] | None = None,
        required: Mapping[str, Sequence[Edge]] | None = None,
    ) -> None:
        self._protocol_name: str = protocol_name
        self._protocol_handle: WasmHandle = protocol_handle
        self._wasm: WasmModule = wasm
        self._ops: tuple[SchemaOp, ...] = tuple(ops) if ops else ()
        self._vertices: dict[str, Vertex] = dict(vertices) if vertices else {}
        self._edges: tuple[Edge, ...] = tuple(edges) if edges else ()
        self._hyper_edges: dict[str, HyperEdge] = dict(hyper_edges) if hyper_edges else {}
        self._constraints: dict[str, tuple[Constraint, ...]] = (
            {k: tuple(v) for k, v in constraints.items()} if constraints else {}
        )
        self._required: dict[str, tuple[Edge, ...]] = (
            {k: tuple(v) for k, v in required.items()} if required else {}
        )

    # ------------------------------------------------------------------
    # Mutation helpers (all return new instances)
    # ------------------------------------------------------------------

    def vertex(
        self,
        id: str,  # noqa: A002
        kind: str,
        options: VertexOptions | None = None,
    ) -> SchemaBuilder:
        """Add a vertex to the schema.

        Parameters
        ----------
        id : str
            Unique vertex identifier within this schema.
        kind : str
            Vertex kind (e.g. ``"record"``, ``"object"``, ``"string"``).
        options : VertexOptions, optional
            Additional vertex configuration such as ``nsid``.

        Returns
        -------
        SchemaBuilder
            A new builder with the vertex added.

        Raises
        ------
        SchemaValidationError
            If *id* is already used by another vertex in this builder.
        """
        if id in self._vertices:
            raise SchemaValidationError(
                f'Vertex "{id}" already exists in schema',
                (f"Duplicate vertex id: {id}",),
            )

        nsid = options.get("nsid") if options else None
        new_vertex: Vertex = {"id": id, "kind": kind, "nsid": nsid}
        op: SchemaOp = {"op": "vertex", "id": id, "kind": kind, "nsid": nsid}

        new_vertices = {**self._vertices, id: new_vertex}

        return SchemaBuilder(
            self._protocol_name,
            self._protocol_handle,
            self._wasm,
            (*self._ops, op),
            new_vertices,
            self._edges,
            self._hyper_edges,
            self._constraints,
            self._required,
        )

    def edge(
        self,
        src: str,
        tgt: str,
        kind: str,
        options: EdgeOptions | None = None,
    ) -> SchemaBuilder:
        """Add a directed edge to the schema.

        Parameters
        ----------
        src : str
            Source vertex id (must already be present in this builder).
        tgt : str
            Target vertex id (must already be present in this builder).
        kind : str
            Edge kind (e.g. ``"record-schema"``, ``"prop"``).
        options : EdgeOptions, optional
            Additional edge configuration such as ``name``.

        Returns
        -------
        SchemaBuilder
            A new builder with the edge added.

        Raises
        ------
        SchemaValidationError
            If *src* or *tgt* are not present in this builder.
        """
        if src not in self._vertices:
            raise SchemaValidationError(
                f'Edge source "{src}" does not exist',
                (f"Unknown source vertex: {src}",),
            )
        if tgt not in self._vertices:
            raise SchemaValidationError(
                f'Edge target "{tgt}" does not exist',
                (f"Unknown target vertex: {tgt}",),
            )

        name = options.get("name") if options else None
        new_edge: Edge = {"src": src, "tgt": tgt, "kind": kind, "name": name}
        op: SchemaOp = {
            "op": "edge",
            "src": src,
            "tgt": tgt,
            "kind": kind,
            "name": name,
        }

        return SchemaBuilder(
            self._protocol_name,
            self._protocol_handle,
            self._wasm,
            (*self._ops, op),
            self._vertices,
            (*self._edges, new_edge),
            self._hyper_edges,
            self._constraints,
            self._required,
        )

    def hyper_edge(
        self,
        id: str,  # noqa: A002
        kind: str,
        signature: Mapping[str, str],
        parent_label: str,
    ) -> SchemaBuilder:
        """Add a hyperedge to the schema.

        Only valid when the protocol's schema theory includes
        ``ThHypergraph``.

        Parameters
        ----------
        id : str
            Unique hyperedge identifier.
        kind : str
            Hyperedge kind.
        signature : Mapping[str, str]
            Label-to-vertex mapping defining the hyperedge signature.
        parent_label : str
            The label in *signature* that identifies the parent vertex.

        Returns
        -------
        SchemaBuilder
            A new builder with the hyperedge added.
        """
        new_he: HyperEdge = {
            "id": id,
            "kind": kind,
            "signature": dict(signature),
            "parent_label": parent_label,
        }
        op: SchemaOp = {
            "op": "hyper_edge",
            "id": id,
            "kind": kind,
            "signature": dict(signature),
            "parent": parent_label,
        }
        new_hyper_edges = {**self._hyper_edges, id: new_he}

        return SchemaBuilder(
            self._protocol_name,
            self._protocol_handle,
            self._wasm,
            (*self._ops, op),
            self._vertices,
            self._edges,
            new_hyper_edges,
            self._constraints,
            self._required,
        )

    def constraint(
        self,
        vertex_id: str,
        sort: str,
        value: str,
    ) -> SchemaBuilder:
        """Add a constraint to a vertex.

        Parameters
        ----------
        vertex_id : str
            The vertex to constrain.
        sort : str
            Constraint sort (e.g. ``"maxLength"``).
        value : str
            Constraint value.

        Returns
        -------
        SchemaBuilder
            A new builder with the constraint added.
        """
        new_c: Constraint = {"sort": sort, "value": value}
        op: SchemaOp = {
            "op": "constraint",
            "vertex": vertex_id,
            "sort": sort,
            "value": value,
        }
        existing = self._constraints.get(vertex_id, ())
        new_constraints = {**self._constraints, vertex_id: (*existing, new_c)}

        return SchemaBuilder(
            self._protocol_name,
            self._protocol_handle,
            self._wasm,
            (*self._ops, op),
            self._vertices,
            self._edges,
            self._hyper_edges,
            new_constraints,
            self._required,
        )

    def required(
        self,
        vertex_id: str,
        edges: Sequence[Edge],
    ) -> SchemaBuilder:
        """Mark edges as required for a vertex.

        Parameters
        ----------
        vertex_id : str
            The vertex whose required edges are being declared.
        edges : Sequence[Edge]
            The edges that must be present for an instance of *vertex_id*.

        Returns
        -------
        SchemaBuilder
            A new builder with the requirement added.
        """
        op: SchemaOp = {
            "op": "required",
            "vertex": vertex_id,
            "edges": [
                {
                    "src": e["src"],
                    "tgt": e["tgt"],
                    "kind": e["kind"],
                    "name": e.get("name"),
                }
                for e in edges
            ],
        }
        existing = self._required.get(vertex_id, ())
        new_required = {**self._required, vertex_id: (*existing, *edges)}

        return SchemaBuilder(
            self._protocol_name,
            self._protocol_handle,
            self._wasm,
            (*self._ops, op),
            self._vertices,
            self._edges,
            self._hyper_edges,
            self._constraints,
            new_required,
        )

    def build(self) -> BuiltSchema:
        """Validate and build the schema.

        Sends all accumulated operations to WASM for validation and
        construction. Returns a :class:`BuiltSchema` that holds both
        the WASM handle and the local data snapshot.

        Returns
        -------
        BuiltSchema
            The validated, built schema.

        Raises
        ------
        SchemaValidationError
            If WASM rejects the schema (invalid structure, missing
            vertices, constraint violations, etc.).
        WasmError
            If the WASM call itself fails unexpectedly.
        """
        ops_bytes = pack_schema_ops(list(self._ops))
        raw_handle = self._wasm.build_schema(
            self._protocol_handle.id,
            ops_bytes,
        )
        handle = create_handle(raw_handle, self._wasm)

        data: SchemaData = {
            "protocol": self._protocol_name,
            "vertices": dict(self._vertices),
            "edges": list(self._edges),
            "hyper_edges": dict(self._hyper_edges),
            "constraints": {k: list(v) for k, v in self._constraints.items()},
            "required": {k: list(v) for k, v in self._required.items()},
        }

        return BuiltSchema(handle, data, self._wasm)


# ---------------------------------------------------------------------------
# BuiltSchema
# ---------------------------------------------------------------------------


@final
class BuiltSchema:
    """A validated, built schema with a WASM-side handle.

    Implements the context-manager protocol for automatic cleanup of
    the underlying WASM resource.

    Parameters
    ----------
    handle : WasmHandle
        The WASM handle returned by ``build_schema``.
    data : SchemaData
        Local snapshot of the schema structure (vertices, edges, etc.).
    wasm : WasmModule
        The WASM module that owns this schema.

    Examples
    --------
    >>> with built_schema:
    ...     migration = panproto.migration(built_schema, other_schema)
    """

    __slots__ = ("_data", "_handle", "_wasm")

    def __init__(
        self,
        handle: WasmHandle,
        data: SchemaData,
        wasm: WasmModule,
    ) -> None:
        self._handle: WasmHandle = handle
        self._data: SchemaData = data
        self._wasm: WasmModule = wasm

    # ------------------------------------------------------------------
    # Internal accessors (used by migration / protocol layers)
    # ------------------------------------------------------------------

    @property
    def wasm_handle(self) -> WasmHandle:
        """The underlying WASM handle (internal use only)."""
        return self._handle

    @property
    def _wasm_module(self) -> WasmModule:
        """The owning WASM module (internal use only)."""
        return self._wasm

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    @property
    def data(self) -> SchemaData:
        """The full schema data snapshot.

        Returns
        -------
        SchemaData
            Immutable snapshot of the schema structure.
        """
        return self._data

    @property
    def protocol(self) -> str:
        """The name of the owning protocol.

        Returns
        -------
        str
        """
        return self._data["protocol"]

    @property
    def vertices(self) -> Mapping[str, Vertex]:
        """All vertices in this schema, keyed by id.

        Returns
        -------
        Mapping[str, Vertex]
        """
        return self._data["vertices"]

    @property
    def edges(self) -> Sequence[Edge]:
        """All directed edges in this schema.

        Returns
        -------
        Sequence[Edge]
        """
        return self._data["edges"]

    # ------------------------------------------------------------------
    # Context manager / cleanup
    # ------------------------------------------------------------------

    def close(self) -> None:
        """Release the WASM-side schema resource.

        Safe to call multiple times; subsequent calls are no-ops.
        """
        self._handle.close()

    def __enter__(self) -> BuiltSchema:
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

File: bindings/typescript/src/schema.ts
/**
 * Fluent schema builder API.
 *
 * Builders are immutable: each method returns a new builder instance.
 * Call `.build()` to validate and produce the final schema.
 *
 * @module
 */

import type {
  WasmModule,
  Vertex,
  Edge,
  HyperEdge,
  Constraint,
  VertexOptions,
  EdgeOptions,
  SchemaData,
  SchemaValidationIssue,
} from './types.js';
import { SchemaValidationError } from './types.js';
import { WasmHandle, createHandle } from './wasm.js';
import type { SchemaOp } from './msgpack.js';
import { packSchemaOps, unpackFromWasm } from './msgpack.js';
import type { Protocol } from './protocol.js';
import { ValidationResult } from './check.js';

/**
 * Immutable fluent builder for constructing schemas.
 *
 * Each mutation method returns a new `SchemaBuilder` instance,
 * leaving the original unchanged. The builder accumulates operations
 * that are sent to WASM on `.build()`.
 *
 * @example
 * ```typescript
 * const schema = builder
 *   .vertex('post', 'record', { nsid: 'app.bsky.feed.post' })
 *   .vertex('post:body', 'object')
 *   .edge('post', 'post:body', 'record-schema')
 *   .build();
 * ```
 */
export class SchemaBuilder {
  readonly #protocolName: string;
  readonly #protocolHandle: WasmHandle;
  readonly #wasm: WasmModule;
  readonly #ops: readonly SchemaOp[];
  readonly #vertices: ReadonlyMap<string, Vertex>;
  readonly #edges: readonly Edge[];
  readonly #hyperEdges: ReadonlyMap<string, HyperEdge>;
  readonly #constraints: ReadonlyMap<string, readonly Constraint[]>;
  readonly #required: ReadonlyMap<string, readonly Edge[]>;

  constructor(
    protocolName: string,
    protocolHandle: WasmHandle,
    wasm: WasmModule,
    ops: readonly SchemaOp[] = [],
    vertices: ReadonlyMap<string, Vertex> = new Map(),
    edges: readonly Edge[] = [],
    hyperEdges: ReadonlyMap<string, HyperEdge> = new Map(),
    constraints: ReadonlyMap<string, readonly Constraint[]> = new Map(),
    required: ReadonlyMap<string, readonly Edge[]> = new Map(),
  ) {
    this.#protocolName = protocolName;
    this.#protocolHandle = protocolHandle;
    this.#wasm = wasm;
    this.#ops = ops;
    this.#vertices = vertices;
    this.#edges = edges;
    this.#hyperEdges = hyperEdges;
    this.#constraints = constraints;
    this.#required = required;
  }

  /**
   * Add a vertex to the schema.
   *
   * @param id - Unique vertex identifier
   * @param kind - Vertex kind (e.g., 'record', 'object', 'string')
   * @param options - Optional vertex configuration (nsid, etc.)
   * @returns A new builder with the vertex added
   * @throws {@link SchemaValidationError} if vertex id is already in use
   */
  vertex(id: string, kind: string, options?: VertexOptions): SchemaBuilder {
    if (this.#vertices.has(id)) {
      throw new SchemaValidationError(
        `Vertex "${id}" already exists in schema`,
        [`Duplicate vertex id: ${id}`],
      );
    }

    const vertex: Vertex = {
      id,
      kind,
      nsid: options?.nsid,
    };

    const op: SchemaOp = {
      op: 'vertex',
      id,
      kind,
      nsid: options?.nsid ?? null,
    };

    const newVertices = new Map(this.#vertices);
    newVertices.set(id, vertex);

    return new SchemaBuilder(
      this.#protocolName,
      this.#protocolHandle,
      this.#wasm,
      [...this.#ops, op],
      newVertices,
      this.#edges,
      this.#hyperEdges,
      this.#constraints,
      this.#required,
    );
  }

  /**
   * Add a directed edge to the schema.
   *
   * @param src - Source vertex id
   * @param tgt - Target vertex id
   * @param kind - Edge kind (e.g., 'record-schema', 'prop')
   * @param options - Optional edge configuration (name, etc.)
   * @returns A new builder with the edge added
   * @throws {@link SchemaValidationError} if source or target vertex does not exist
   */
  edge(src: string, tgt: string, kind: string, options?: EdgeOptions): SchemaBuilder {
    if (!this.#vertices.has(src)) {
      throw new SchemaValidationError(
        `Edge source "${src}" does not exist`,
        [`Unknown source vertex: ${src}`],
      );
    }
    if (!this.#vertices.has(tgt)) {
      throw new SchemaValidationError(
        `Edge target "${tgt}" does not exist`,
        [`Unknown target vertex: ${tgt}`],
      );
    }

    const edge: Edge = {
      src,
      tgt,
      kind,
      name: options?.name,
    };

    const op: SchemaOp = {
      op: 'edge',
      src,
      tgt,
      kind,
      name: options?.name ?? null,
    };

    return new SchemaBuilder(
      this.#protocolName,
      this.#protocolHandle,
      this.#wasm,
      [...this.#ops, op],
      this.#vertices,
      [...this.#edges, edge],
      this.#hyperEdges,
      this.#constraints,
      this.#required,
    );
  }

  /**
   * Add a hyperedge to the schema.
   *
   * Only valid if the protocol's schema theory includes ThHypergraph.
   *
   * @param id - Unique hyperedge identifier
   * @param kind - Hyperedge kind
   * @param signature - Label-to-vertex mapping
   * @param parentLabel - The label identifying the parent in the signature
   * @returns A new builder with the hyperedge added
   */
  hyperEdge(
    id: string,
    kind: string,
    signature: Record<string, string>,
    parentLabel: string,
  ): SchemaBuilder {
    const he: HyperEdge = { id, kind, signature, parentLabel };

    const op: SchemaOp = {
      op: 'hyper_edge',
      id,
      kind,
      signature,
      parent: parentLabel,
    };

    const newHyperEdges = new Map(this.#hyperEdges);
    newHyperEdges.set(id, he);

    return new SchemaBuilder(
      this.#protocolName,
      this.#protocolHandle,
      this.#wasm,
      [...this.#ops, op],
      this.#vertices,
      this.#edges,
      newHyperEdges,
      this.#constraints,
      this.#required,
    );
  }

  /**
   * Add a constraint to a vertex.
   *
   * @param vertexId - The vertex to constrain
   * @param sort - Constraint sort (e.g., 'maxLength')
   * @param value - Constraint value
   * @returns A new builder with the constraint added
   */
  constraint(vertexId: string, sort: string, value: string): SchemaBuilder {
    const c: Constraint = { sort, value };

    const op: SchemaOp = {
      op: 'constraint',
      vertex: vertexId,
      sort,
      value,
    };

    const existing = this.#constraints.get(vertexId) ?? [];
    const newConstraints = new Map(this.#constraints);
    newConstraints.set(vertexId, [...existing, c]);

    return new SchemaBuilder(
      this.#protocolName,
      this.#protocolHandle,
      this.#wasm,
      [...this.#ops, op],
      this.#vertices,
      this.#edges,
      this.#hyperEdges,
      newConstraints,
      this.#required,
    );
  }

  /**
   * Mark edges as required for a vertex.
   *
   * @param vertexId - The vertex with required edges
   * @param edges - The edges that are required
   * @returns A new builder with the requirement added
   */
  required(vertexId: string, edges: readonly Edge[]): SchemaBuilder {
    const op: SchemaOp = {
      op: 'required',
      vertex: vertexId,
      edges: edges.map((e) => ({
        src: e.src,
        tgt: e.tgt,
        kind: e.kind,
        name: e.name ?? null,
      })),
    };

    const existing = this.#required.get(vertexId) ?? [];
    const newRequired = new Map(this.#required);
    newRequired.set(vertexId, [...existing, ...edges]);

    return new SchemaBuilder(
      this.#protocolName,
      this.#protocolHandle,
      this.#wasm,
      [...this.#ops, op],
      this.#vertices,
      this.#edges,
      this.#hyperEdges,
      this.#constraints,
      newRequired,
    );
  }

  /**
   * Validate and build the schema.
   *
   * Sends all accumulated operations to WASM for validation and construction.
   * Returns a `BuiltSchema` that holds the WASM handle and local data.
   *
   * @returns The validated, built schema
   * @throws {@link SchemaValidationError} if the schema is invalid
   */
  build(): BuiltSchema {
    const opsBytes = packSchemaOps([...this.#ops]);
    const rawHandle = this.#wasm.exports.build_schema(
      this.#protocolHandle.id,
      opsBytes,
    );

    const handle = createHandle(rawHandle, this.#wasm);

    const data: SchemaData = {
      protocol: this.#protocolName,
      vertices: Object.fromEntries(this.#vertices),
      edges: [...this.#edges],
      hyperEdges: Object.fromEntries(this.#hyperEdges),
      constraints: Object.fromEntries(
        Array.from(this.#constraints.entries()).map(([k, v]) => [k, [...v]]),
      ),
      required: Object.fromEntries(
        Array.from(this.#required.entries()).map(([k, v]) => [k, [...v]]),
      ),
      variants: {},
      orderings: {},
      recursionPoints: {},
      usageModes: {},
      spans: {},
      nominal: {},
    };

    return new BuiltSchema(handle, data, this.#wasm);
  }
}

/**
 * A validated, built schema with a WASM-side handle.
 *
 * Implements `Disposable` for automatic cleanup of the WASM resource.
 */
export class BuiltSchema implements Disposable {
  readonly #handle: WasmHandle;
  readonly #data: SchemaData;
  readonly #wasm: WasmModule;

  constructor(handle: WasmHandle, data: SchemaData, wasm: WasmModule) {
    this.#handle = handle;
    this.#data = data;
    this.#wasm = wasm;
  }

  /** The WASM handle for this schema. Internal use only. */
  get _handle(): WasmHandle {
    return this.#handle;
  }

  /** The WASM module reference. Internal use only. */
  get _wasm(): WasmModule {
    return this.#wasm;
  }

  /** The schema data (vertices, edges, constraints, etc.). */
  get data(): SchemaData {
    return this.#data;
  }

  /** The protocol name this schema belongs to. */
  get protocol(): string {
    return this.#data.protocol;
  }

  /** All vertices in the schema. */
  get vertices(): Readonly<Record<string, Vertex>> {
    return this.#data.vertices;
  }

  /** All edges in the schema. */
  get edges(): readonly Edge[] {
    return this.#data.edges;
  }

  /** @internal Create from raw handle (used by normalize). */
  static _fromHandle(
    handle: number,
    data: SchemaData,
    _protocol: string,
    wasm: WasmModule,
  ): BuiltSchema {
    const wasmHandle = createHandle(handle, wasm);
    return new BuiltSchema(wasmHandle, data, wasm);
  }

  /** Normalize this schema by collapsing reference chains. Returns a new BuiltSchema. */
  normalize(): BuiltSchema {
    const handle = this.#wasm.exports.normalize_schema(this.#handle.id);
    return BuiltSchema._fromHandle(handle, this.#data, this.#data.protocol, this.#wasm);
  }

  /** Validate this schema against a protocol's rules. */
  validate(protocol: Protocol): ValidationResult {
    const bytes = this.#wasm.exports.validate_schema(
      this.#handle.id,
      protocol._handle.id,
    );
    const issues = unpackFromWasm<SchemaValidationIssue[]>(bytes);
    return new ValidationResult(issues);
  }

  /** Release the WASM-side schema resource. */
  [Symbol.dispose](): void {
    this.#handle[Symbol.dispose]();
  }
}
